Source/Details/ASCollectionLayoutCalculating.h

```diff
@@ -0,0 +1,64 @@
+//
+//  ASCollectionLayoutCalculating.h
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 28/2/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <AsyncDisplayKit/ASDimension.h>
+
+@class ASLayout, ASElementMap;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol ASCollectionLayoutCalculating <NSObject>
+
+/**
+ * Whether or not size ranges must be provided via collection element in the element map.
+ * Default to NO, which means the calculator can determine size range for each collection element.
+ * ASLegacyCollectionLayoutCalculator might return YES because size ranges used to be provided via async delegate of table/collection view or layout inspector of collection view.
+ */
+@property (nonatomic, assign, readonly) BOOL sizeRangesShouldBeProvided;
+
+/**
+ * @abstract Return a layout of elements.
+ *
+ * @param viewportSize The viewport size that the resulting layout should fit, usually equals to the bounds of the containing table/collection view.
+ *
+ * @param map An element map containings all elements to be laid out.
+ *
+ * @discussion This method can be called on multiple theads and thus must be thread-safe (and ideally, stateless).
+ * This method must block the calling thread but can dispatch to other theads to reduce the blocking time.
+ */
+- (ASLayout *)layoutThatFits:(CGSize)viewportSize forElementMap:(ASElementMap *)map;
```

Comment:
Also, should we pass some kind of context/helper object?

For example, we could remove the two arguments here, and the `sizeRangesShouldBeProvided` flag in the protocol if we gave them something like:

```objc
ASCollectionLayoutContext {
   CGSize viewportSize;
   id<ASCollectionDelegate> collectionDelegate;
   ASElementMap *map;
   - (void)parallelAllocateNodesForElements:(id<NSFastEnumeration> elements);
    // Maybe?
   - (void)setFrame:(CGRect)frame forElement:(ASCollectionElement *)element;
   - (BOOL)isCancelled;
}
```